### ISSUE TITLE
kabanero operator reconcile problem if namespace is created after rolebinding

### ISSUE THREAD

**Issue by tam512** [kabanero-io/kabanero-operator#702]
Use Kabanero 0.9.0-rc.2, run pipeline nodejs-express-build-deploy-pl in Tekton dashboard to deploy an application to svt-project2 namespace.   I added svt-project2 to `spec.targetNamespaces` in Kabanero CR, but I forgot to create svt-project2 the namespace so the deploy-image step failed.
So I created the svt-project2 namespace and reran the pipeline, but it still failed because the namespace was created afterwards, so it didn't loop back around to create the rolebinding for permission again.

The current workaround is to restart kabanero operator pod (_oc -n kabanero delete pod -l name=kabanero-operator_)

Below is the error on deploy-image step
```
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "appsody.dev/v1beta1, Resource=appsodyapplications", GroupVersionKind: "appsody.dev/v1beta1, Kind=AppsodyApplication"
Name: "radcl-njsexpress-app", Namespace: "svt-project2"
Object: &{map["apiVersion":"appsody.dev/v1beta1" "kind":"AppsodyApplication" "metadata":map["annotations":map["architecture":"x86_64" "authoritative-source-url":"registry.access.redhat.com" "build-date":"2020-04-09T10:02:05.547594" "commit.image.appsody.dev/author":"Tam T Dinh <tamdinh@us.ibm.com>" "commit.image.appsody.dev/committer":"GitHub Enterprise <noreply@github.ibm.com>" "commit.image.appsody.dev/date":"Mon May 11 13:09:41 2020 -0400" "commit.image.appsody.dev/message":"Update app-deploy.yaml" "commit.stack.appsody.dev/contextDir":"/incubator/nodejs-express" "commit.stack.appsody.dev/date":"Fri May 1 15:13:54 2020 +0100" "commit.stack.appsody.dev/message":"Merge pull request #331 from uberskigeek/329_fix_missing_continuation" "description":"This image contains the Kabanero development stack for the Nodejs Express collection" "distribution-scope":"public" "help":"For more information visit https://github.com/sclorg/s2i-nodejs-container" "image.opencontainers.org/created":"2020-05-11T17:55:45Z" "image.opencontainers.org/documentation":"https://github.ibm.com/tamdinh/radcl-njsexpress-app" "image.opencontainers.org/revision":"2e5c844aa83988d40759e63071d5dbf7685af45d" "image.opencontainers.org/source":"https://github.ibm.com/tamdinh/radcl-njsexpress-app/tree/master" "image.opencontainers.org/url":"https://github.ibm.com/tamdinh/radcl-njsexpress-app" "k8s.io/description":"Node.js 12 available as container is a base platform for building and running various Node.js 12 applications and frameworks. Node.js is a platform built on Chrome's JavaScript runtime for easily building fast, scalable network applications. Node.js uses an event-driven, non-blocking I/O model that makes it lightweight and efficient, perfect for data-intensive real-time applications that run across distributed devices." "k8s.io/display-name":"Node.js 12" "kubectl.kubernetes.io/last-applied-configuration":"" "maintainer":"SoftwareCollections.org <sclorg@redhat.com>" "name":"kabanero/nodejs-express" "openshift.io/expose-services":"8080:http" "openshift.io/tags":"builder,nodejs,nodejs12" "redhat.com/build-host":"cpt-1004.osbs.prod.upshift.rdu2.redhat.com" "redhat.com/component":"nodejs-12-container" "redhat.com/deployments-dir":"/opt/app-root/src" "redhat.com/dev-mode":"DEV_MODE:false" "redhat.com/dev-mode.port":"DEBUG_PORT:5858" "redhat.com/license_terms":"https://www.redhat.com/en/about/red-hat-end-user-license-agreements#UBI" "release":"36" "s2i.io/scripts-url":"image:///usr/libexec/s2i" "s2i.openshift.io/scripts-url":"image:///usr/libexec/s2i" "stack.appsody.dev/authors":"Sam Roberts <sam-github>" "stack.appsody.dev/configured":"docker.io/kabanerobeta/nodejs-express:0.4" "stack.appsody.dev/created":"2020-05-01T14:56:29Z" "stack.appsody.dev/description":"Express web framework for Node.js" "stack.appsody.dev/digest":"sha256:95cd77cb2f7eb5b336b430c8bf775367c38ee200c74f73fe73b02c61fff932f9" "stack.appsody.dev/documentation":"https://github.com/kabanero-io/collections/tree/master/incubator/nodejs-express/README.md" "stack.appsody.dev/licenses":"Apache-2.0" "stack.appsody.dev/revision":"8c75520a677c818df6c9c546b9402cc1733feaa6" "stack.appsody.dev/source":"https://github.com/kabanero-io/collections/tree/master/incubator/nodejs-express/image" "stack.appsody.dev/tag":"docker.io/kabanerobeta/nodejs-express:0.4.8" "stack.appsody.dev/title":"Node.js Express" "stack.appsody.dev/url":"https://github.com/kabanero-io/collections/tree/master/incubator/nodejs-express" "summary":"Image for Kabanero Node.js Express development" "url":"https://access.redhat.com/containers/#/registry.access.redhat.com/ubi8/nodejs-12/images/1-36" "usage":"s2i build <SOURCE-REPOSITORY> ubi8/nodejs-12:latest <APP-NAME>" "vcs-ref":"9a85b86edcc6c765692f15311abe259b0f9997b2" "vcs-type":"git" "vendor":"Kabanero" "version":"0.4.8"] "creationTimestamp":<nil> "labels":map["image.opencontainers.org/title":"radcl-njsexpress-app" "stack.appsody.dev/id":"nodejs-express" "stack.appsody.dev/version":"0.4.8"] "name":"radcl-njsexpress-app" "namespace":"svt-project2"] "spec":map["applicationImage":"index.docker.io/tamdocker/radcl-njsexpress-app:2e5c844aa83988d40759e63071d5dbf7685af45d" "createKnativeService":%!q(bool=false) "expose":%!q(bool=true) "livenessProbe":map["failureThreshold":'\f' "httpGet":map["path":"/live" "port":'\u0bb8'] "initialDelaySeconds":'\x05' "periodSeconds":'\x02'] "monitoring":map["labels":map["k8s-app":"radcl-njsexpress-app"]] "readinessProbe":map["failureThreshold":'\f' "httpGet":map["path":"/ready" "port":'\u0bb8'] "initialDelaySeconds":'\x05' "periodSeconds":'\x02' "timeoutSeconds":'\x01'] "service":map["annotations":map["prometheus.io/scrape":"true"] "port":'\u0bb8' "type":"NodePort"] "stack":"nodejs-express" "version":"1.0.0"]]}
from server for: "/workspace/git-source/app-deploy.yaml": appsodyapplications.appsody.dev "radcl-njsexpress-app" is forbidden: User "system:serviceaccount:kabanero:kabanero-pipeline" cannot get resource "appsodyapplications" in API group "appsody.dev" in the namespace "svt-project2"
```

**Comment by kaczyns** [comment]
Some options to fix this:
1) Watch on namespace.  I'm not sure a namespace scoped operator can do this (watch a cluster scoped object).  If we can then we can watch for a namespace with that name to be created, and then create the role / apply the role binding.
2) Let the kabanero reconciler get into a loop if it detects the namespace does not exist.  It can treat it as an error and just keep retrying until either the namespace exists, or the kabanero CR codes a new namespace.

As I write this, 2 is probably the easier option.

**Comment by kaczyns** [comment]
There is some function being added to the CLI service which is going to require the CLI service to scan for deployments in `targetNamespaces`.  This means that the CLI service SA is also going to have selective role bindings based (in addition to the pipelines SA).  The same problems that Tam is hitting are going to apply to CLI service as well.  I think we need a framework around this because I doubt that these are going to be the only two cases where some namespace-scoped Kabanero component needs to reach out into another namespace.

I think that the Kabanero CR instance status adopts a "start over" approach to status whenever the reconciler runs.  This may need to change.  I think the reconciler needs to do something like this (if we're not doing it already - I apologize for not looking in master to check):
1) Compare the `spec.targetNamespaces` against the `status.targetNamespaces`.
2) For new namespaces in `spec.targetNamespaces`, make sure the namespace exists.  If it does not, error out without changing `status.targetNamespaces`.
3) For namespaces in common between `spec.targetNamespaces` and `status.targetNamespaces`, just check to make sure the role binding is set correctly.
4) For namespaces in `status.targetNamespaces` but not in `spec.targetNamespaces`, delete the role binding.
5) For namespaces in `spec.targetNamespaces` but not in `status.targetNamespaces`, create the role binding.  This may end up being the same as 3).
6) Update the status so that `status.targetNamespaces` == `spec.targetNamespaces`

If we do it this way, the status will reflect what the current permissions are.  If there is some error about a namespace not existing, it will be clear that the reconciler still has work to do.  We're less likely to leave things in a half-created state.  Once all the namespaces exist we can create the bindings, and move on.

The target namespaces should probably get its own `.go` part (like the cli and landing etc do) which processes all the role bindings that come from the target namespaces.

**Comment by marikaj123** [comment]
@tam512 - As was suggested, the issue is nice to have in Kabanero 0.9.0.

**Comment by kaczyns** [comment]
This was fixed in #717
